sys/src/9/amd64/devvq.c

```diff
@@ -0,0 +1,595 @@
+	default:
+		return -1;
+	}
+	return -1;
+}
+
+
+static Chan*
+vqattach(char *spec)
+{
+//print("vqattach %s\n", spec);
+	return devattach(vqdevtab.dc, spec);
+}
+
+
+Walkqid*
+vqwalk(Chan* c, Chan *nc, char** name, int nname)
+{
+//print("vqwalk %d -> %d\n", c?TYPE(c->qid):-1, nc?TYPE(nc->qid):-1);
+	return devwalk(c, nc, name, nname, (Dirtab *)0, 0, vqgen);
+}
+
+static int32_t
+vqstat(Chan* c, uint8_t* dp, int32_t n)
+{
+//print("vqstat %d\n", TYPE(c->qid));
+	return devstat(c, dp, n, (Dirtab *)0, 0L, vqgen);
+}
+
+// A process holding open the device configuration area has access to its virtqueues.
+// Other processes will be denied any access. PID of the claiming process is stored
+// in the ownpid field of the device descriptor. An unclaimed device has -1 in it.
+
+static Chan*
+vqopen(Chan *c, int omode)
+{
+//print("v9open %d\n", TYPE(c->qid));
+	Proc *up = externup();
+	uint t = TYPE(c->qid);
+	uint vdidx = DEV(c->qid);
+	if(vdidx >= nvq)
+		error(Ebadarg);
+	vqctl *vc = cvq[vdidx];
+	switch(t) {
+		case Qdevcfg:
+			if(vc->ownpid != -1)
+				error(Eperm);
+			vc->ownpid = up->pid;
+		case Qvqdata:
+		case Qvqctl:
+			if(vc->ownpid != up->pid)
+				error(Eperm);
+			break;
+		default:
+			break;
+	}
+	c = devopen(c, omode, (Dirtab*)0, 0, vqgen);
+	switch(t) {
+	default:
+		break;
+	}
+	return c;
+}
+
+static void
+vqclose(Chan* c)
+{
+//print("vqclose %d\n", TYPE(c->qid));
+	Proc *up = externup();
+	uint t = TYPE(c->qid);
+	uint vdidx = DEV(c->qid);
+	if(vdidx >= nvq)
+		return;
+	vqctl *vc = cvq[vdidx];
+	if (vc->ownpid != up->pid)
+		return;
+	switch(t) {
+		case Qdevcfg:
+			vc->ownpid = -1;
+			break;
+		default:
+			break;
+	}
+}
+
+
+
+// Reading from the device configuration area is byte-by-byte (8bit ports) to avoid dealing with
+// endianness.
+
+static int32_t
+vqread(Chan *c, void *va, int32_t n, int64_t offset)
+{
+//print("vqread %d %d %d\n", TYPE(c->qid), n, offset);
+	int vdidx = DEV(c->qid);
+	int8_t *a;
+	uint32_t r;
+	switch(TYPE(c->qid)) {
+	case Qtopdir:
+	case Qvirtqs:
+	case Qdevtop:
+		return devdirread(c, va, n, (Dirtab *)0, 0L, vqgen);
+	case Qdevcfg:
+		if(vdidx >= nvq) {
+			error(Ebadarg);
+		}
+		a = va;
+		r = offset;
+		int i;
+		for(i = 0; i < n; a++, i++) {
+			if(i + r >= cvq[vdidx]->dcfglen)
+				break;
+			uint8_t b = inb(cvq[vdidx]->port + cvq[vdidx]->dcfgoff + i + r);
+			PBIT8(a, b);
+		}
+		return i;
+	default:
+		return -1;
+	}
+}
+
+static int32_t
+vqwrite(Chan *c, void *va, int32_t n, int64_t offset)
+{
+//print("vqwrite %d %d %d\n", TYPE(c->qid), n, offset);
+	return -1;
+}
+
+// Scan virtqueues for the given device. If the vqs argument is not nil then
+// nvq is expected to contain the length of the array vqs points to. In this case
+// populate the Virtq structures for each virtqueue found. Otherwise just return
+// the number of virtqueues detected. The port argument contains the base port
+// for the device being scanned.
+// Portions of code are borrowed from the 9front virtio drivers.
```

Comment:
OK we can't just take code that is "borrowed". We have to manage the license correctly. Did you talk to cinap about the license question? He's been very nice about releasing his code to us under MIT or BSD license.